lucas_scheinkerman/Debugging_and_profiling_in_ROS.md

```diff
@@ -0,0 +1,144 @@
+
+```console
+(gdb) break 12
+Breakpoint 1 at 0x47e5f4: file /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp, line 12.
+(gdb) break 16
+Breakpoint 2 at 0x47e623: file /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp, line 16.
+(gdb) break 18
+Breakpoint 3 at 0x47e689: file /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp, line 18.
+(gdb) 
+```
+
+No problems were printed. Let's run the program:
+
+```console
+(gdb) r
+Starting program: /create_ws/devel/lib/ca_move_base/send_goal 1 1 2 __name:=nav __log:=/home/create/.ros/log/ba42386a-da4a-11e9-81ae-902b346723a7/create1-nav-2.log
+[Thread debugging (gdb) r
+Starting program: /create_ws/devel/lib/ca_move_base/send_goal 1 1 2 __name:=nav __log:=/home/create/.ros/log/ba42386a-da4a-11e9-81ae-902b346723a7/create1-nav-2.log
+[Thread debugging using libthread_db enabled]
+Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
+
+Breakpoint 1, main (argc=6, argv=0x7fffffffe178) at /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp:12
+12      int main(int argc, char** argv){
+(gdb) next
+14              ros::init(argc, argv, NODE_NAME);
+(gdb) next
+
+Breakpoint 2, main (argc=4, argv=0x7fffffffe178) at /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp:16
+16              MoveBaseClient ac("move_base", true);
+(gdb) next
+[New Thread 0x7ffff1636700 (LWP 6886)]
+[New Thread 0x7ffff0e35700 (LWP 6887)]
+[New Thread 0x7fffebfff700 (LWP 6888)]
+[New Thread 0x7fffeb7fe700 (LWP 6893)]
+[New Thread 0x7fffeaffd700 (LWP 6894)]
+
+Thread 1 "send_goal" hit Breakpoint 3, main (argc=4, argv=0x7fffffffe178) at /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp:18
+18              while(!ac.waitForServer(ros::Duration(5.0))){using libthread_db enabled]
+Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
+
+Breakpoint 1, main (argc=6, argv=0x7fffffffe178) at /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp:12
+12      int main(int argc, char** argv){
+(gdb) next
+14              ros::init(argc, argv, NODE_NAME);
+(gdb) next
+
+Breakpoint 2, main (argc=4, argv=0x7fffffffe178) at /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp:16
+16              MoveBaseClient ac("move_base", true);
+(gdb) next
+[New Thread 0x7ffff1636700 (LWP 6886)]
+[New Thread 0x7ffff0e35700 (LWP 6887)]
+[New Thread 0x7fffebfff700 (LWP 6888)]
+[New Thread 0x7fffeb7fe700 (LWP 6893)]
+[New Thread 0x7fffeaffd700 (LWP 6894)]
+
+Thread 1 "send_goal" hit Breakpoint 3, main (argc=4, argv=0x7fffffffe178) at /create_ws/src/create_autonomy/navigation/ca_move_base/src/send_robot_goal.cpp:18
+18              while(!ac.waitForServer(ros::Duration(5.0))){
+```
+
+It can be seen that the `.cpp` program was executed without any issues regarding to the use of the `gdb` program.
```

Comment:
Could you run another example than `breakpoint` in `gdb` just to understand its utility? What else can you do with it?